### PR TITLE
Smoothing

### PR DESCRIPTION
**Pull request by ismailsunni**
### What does it fix?
* Ticket: #4441 
* Funded by: DFAT
* Description: 
   - Adding smoothing with scipy (from realtime) and with numpy
   - Add smoothing option in the shake map converter dialog
      screen shot 2017-11-13 at 12 02 12
   - Scipy option will be disable if there is no scipy installed
   - The result using numpy and scipy is the same 
     screen shot 2017-11-13 at 12 04 55
   - Small improvement in shake map converter dialog (open file behaviour)

### Checklist:
- [x] Commit messages are descriptive and explain the rationale for changes
- [ ] Add to the changelog in metadata.txt if it's a new feature
- [ ] Unit test for new code added
- [x] Request someone to review or test your PR

### CODE REVIEW

File: safe/gui/tools/shake_grid/shakemap_converter_dialog.py [hunk 1]
```diff
@@ -238,28 +268,37 @@ def accept(self):
     def on_open_input_tool_clicked(self):
         """Autoconnect slot activated when open input tool button is clicked.
         """
+        input_path = self.input_path.text()
+        if not input_path:
+            input_path = os.path.expanduser('~')
```
Comment (Gustry) [inline safe/gui/tools/shake_grid/shakemap_converter_dialog.py]: Something that would be very nice to fix for the user experience, it's to follow the last input directory. We should use the QSettings in the `UI` section.
screen shot 2017-11-13 at 07 43 28

I know a lot of users hate when a software doesn't remember your last used folder and bring you to your home directory everytime. Something we should think of when we create new dialog.

File: safe/gui/tools/shake_grid/shakemap_converter_dialog.py [hunk 2]
```diff
@@ -99,7 +120,7 @@ def on_output_path_textChanged(self):
     def on_input_path_textChanged(self):
         """Action when input file name is changed."""
         input_path = self.input_path.text()
-        input_not_grid_msg = self.tr('input file is not .xml')
+        input_not_grid_msg = tr('input file is not .xml')
```
Comment (Gustry) [inline safe/gui/tools/shake_grid/shakemap_converter_dialog.py]: Should we do a search and replace in the code for `self.tr(` to `tr(` ?

**Comment by Gustry** [comment]
I'm pretty surprised that shasum is giving the same result. One pixel different and the shasum must be different. But using two different python functions, with rounding etc ... Whaou.

So does it mean we don't need the pre-processor thing I created last week? :D

File: safe/gui/tools/shake_grid/shakemap_converter_dialog.py (hunk 2 above)
Comment (ismailsunni) [reply safe/gui/tools/shake_grid/shakemap_converter_dialog.py]: Yeah, I think we can do that. It should be easy I think.

File: safe/gui/tools/shake_grid/shakemap_converter_dialog.py (hunk 1 above)
Comment (ismailsunni) [reply safe/gui/tools/shake_grid/shakemap_converter_dialog.py]: Yes, I also hate it actually.
Is there a method or widget that do it automatically? Like checking `last` + `widget name or something` in QGIS?

**Comment by ismailsunni** [comment]
> I'm pretty surprised that shasum is giving the same result. One pixel different and the shasum must be different. But using two different python functions, with rounding etc ... Whaou.

Yeah, me too. It's exactly the same. Actually, I just wanted to compare the result by checking the looks, but it turns out returning the same file.

> So does it mean we don't need the pre-processor thing I created last week? :D

Why? I thought we need it to make run automatically, don't we?

File: safe/definitions/constants.py [hunk 3]
```diff
@@ -77,3 +77,9 @@
 
 # Layer properties
 MULTI_EXPOSURE_ANALYSIS_FLAG = 'multi_exposure_analysis'
+
+
+# Smoothing mode
+NONE_SMOOTHING = 'none_smoothing'
+NUMPY_SMOOTHING = 'numpy_smoothing'
+SCIPY_SMOOTHING = 'scipy_smoothing'
```
Comment (Gustry) [inline safe/definitions/constants.py]: Quick note, just noticed this weekend that Python as a Enum function: https://docs.python.org/3/library/enum.html
It has been backported to Python 2.7.

**Comment by ismailsunni** [comment]
Btw, the one that use numpy is much slower than the scipy one.
```
2017-11-13T15:50:08	0	We are using NUMPY smoothing
2017-11-13T15:50:08	0	Duration : 0.325406
2017-11-13T15:50:51	0	We are using SCIPY smoothing
2017-11-13T15:50:51	0	Duration : 0.01058
```

It's like 30x slower. Is it fast enough @lucernae ?

File: safe/gui/tools/shake_grid/shakemap_converter_dialog.py (hunk 1 above)
Comment (Gustry) [reply safe/gui/tools/shake_grid/shakemap_converter_dialog.py]: QgsFileWidget, but we still need to load/save the last used path.

File: safe/definitions/constants.py (hunk 3 above)
Comment (ismailsunni) [reply safe/definitions/constants.py]: ah nice, we should use it in InaSAFE

**Comment by Gustry** [comment]
> Why? I thought we need it to make run automatically, don't we?

Is it creating the contour file already?
We don't want to create it twice

**Comment by ismailsunni** [comment]
> Is it creating the contour file already?
We don't want to create it twice

Yes. In the realtime, we create two shake map. One without smoothing, and one with smoothing. The one with smoothing is used for displaying the map only.
from realtime code:
```
        # Convert grid.xml (we'll give the title with event_id)
        # RM: convert event_id to str too. This avoid the layer name is
        # falsely read as int
        self.shake_grid = ShakeGrid(
            str(self.event_id), get_grid_source(), self.grid_file_path())
        # RM: add smoothed grid version. Only used for display in the map
        self.shake_grid_smooth = ShakeGrid(
            str(self.event_id), get_grid_source(), self.grid_file_path(),
            output_basename='mmi-smooth',
            smoothed=True)
```

**Comment by Gustry** [comment]
Hum, ok.
So shouldn't we move the code to anoher file?

**Comment by ismailsunni** [comment]
@Gustry I think so. We need to see if the parts the we created (multi exposure, direction, bearing, smoothing) are connected properly so that InaSAFE desktop can do what InaSAFE realtime currently does.

Can we merge this first, then check what is missing to replicate realtime?

**Pull request merged**